### ISSUE TITLE
Feature: Search & Replace

### ISSUE THREAD

**Issue by extr0py** [extr0py/oni#614]
For Oni, I'd like there to be an embedded / native search and replace capability, with the ability to find across files, and replace across files.

This should be easy to navigate, and I should be able to use the existing keybindings and knowledge I have in navigation through neovim/buffers in order to leverage this UI and the results.

This issue is to discuss what sort of functionality and flow we'd like to see in search and replace.

- For my workflows, I'd prefer the flexibility to either have search results on the bottom, or on the side. But not sure exactly how it should look like yet!
- A key tenent across this, and all the new pieces of UI (like file explorer, git integration, find files, errors, etc) - is that I can _navigate using the same paradigms and functionality I use in VIM_ - in other words, I should be able to use h/j/k/l and all the other tools I'm comfortable with. In other words, we should leverage Neovim as the 'engine' for manipulating input, even if we put a fancy or different visualization on top of it.

**Comment by cyansprite** [comment]
> For my workflows, I'd prefer the flexibility to either have search results on the bottom, or on the side. But not sure exactly how it should look like yet!

Right... I'm not sure how it would work either, I've been thinking about this a lot (especially where I was making a plug in) and I got a really good feel for what I like and what I don't.

I've discovered that as far as typing in the search I would want it to be at the bottom of the screen.

Then to open in a vsp, or a new tab (we can make our own dialog even? my plugin just made a scratch buffer with a few options) with the results updating as they find them, I want to preview, split it, vsp it, or tab it with a command (p,s,v,t something like that)

Fold the regions so I can navigate with zj, zk.

There are a few more things... but those are the main things.

Sometime I will demo my plugin and show you how it works, and then see if we can create something similar but of course better with nice graphics and what not.

**Comment by cyansprite** [comment]
Eh... I hate that its fast only if it loads all at once, api calls to update the buffer is slow...
And if i hate python -.-... Hates a strong word but i don't like it.

I say if you are ready to implement find and replace because its something I feel that we need rather soon ( I miss it frequently when using oni )

That way we can also gather files quickly too. Rg can do that and I will work on building something more suited to an ide rather than a binary you use in a command line.

What are your thoughts?
